src/battle_ai_switch_items.c

```diff
@@ -897,37 +1079,774 @@ static u32 GetBestMonDmg(struct Pokemon *party, int firstId, int lastId, u8 inva
     int i, j;
     int dmg, bestDmg = 0;
     int bestMonId = PARTY_SIZE;
+    u32 aiMove;
 
     gMoveResultFlags = 0;
     // If we couldn't find the best mon in terms of typing, find the one that deals most damage.
     for (i = firstId; i < lastId; i++)
     {
         if (gBitTable[i] & invalidMons)
             continue;
-        if (IsAiPartyMonOHKOBy(battler, opposingBattler, &party[i]))
+        InitializeSwitchinCandidate(&party[i]);
+        for (j = 0; j < MAX_MON_MOVES; j++)
+        {
+            if (aiMove != MOVE_NONE && gBattleMoves[aiMove].power != 0)
+            {
+                aiMove = GetMonData(&party[i], MON_DATA_MOVE1 + j);
+                dmg = AI_CalcPartyMonDamage(aiMove, battler, opposingBattler, AI_DATA->switchinCandidate.battlemon, TRUE);
+                if (bestDmg < dmg)
+                {
+                    bestDmg = dmg;
+                    bestMonId = i;
+                }
+            }
+        }
+    }
+
+    return bestMonId;
+}
+
+// Gets hazard damage
+static u32 GetSwitchinHazardsDamage(u32 battler, struct BattlePokemon battlemon)
```

Comment:
The argument should be struct BattlePokemon *battleMon. 
Right now you're passing the whole struct instead of a pointer. It means that whenever this function is called, it creates a local variable on the stack which takes additional time, because it copies the same data twice. 
If you're curious and know a bit of asm, you can have 2 version of this function, one which is the same and one which takes a pointer argument(then accessed as battleMon-> instead of battleMon.), compile, and see the asm output(using make KEEP_TEMPS). If all of that sounds like nonsense, tldr is your version takes more computation time.